core/state/database.go

```diff
@@ -135,15 +135,16 @@ func NewDatabaseWithConfig(db ethdb.Database, config *trie.Config) Database {
 		db:            trie.NewDatabaseWithConfig(db, config),
 		disk:          db,
 		codeSizeCache: csc,
-		codeCache:     fastcache.New(codeCacheSize),
+		//codeCache:     fastcache.New(codeCacheSize),
+		codeCache: lru2.NewSizeConstraiedLRU(codeCacheSize),
 	}
 }
 
 type cachingDB struct {
 	db            *trie.Database
 	disk          ethdb.KeyValueStore
 	codeSizeCache *lru.Cache
-	codeCache     *fastcache.Cache
+	codeCache     *lru2.SizeConstrainedLRU
```

Comment:
The simpleLRU is not thread-safe, but we will in the concurrent way. A proof is in function `ContractCodeWithPrefix` we use it for serving `GetNodeData` protocol requests.

Reply:
We should use the `"github.com/hashicorp/golang-lru"` instead, which is the wrapper of simpleLRU in a thread-safe way.

Reply:
Right.. Unfortunately, we cannot just use the golang-lru `Cache` (lru.go), and let it worry abouth the concurrency - because the `Set` operation which does eviction needs to hold the mutex the entire time. So I'll just add the necessary locks here instead